[Doc] Revamp ray core design patterns doc [11/n]: limit running tasks (#29046)
ray-project/ray: doc/source/ray-core/patterns/limit-running-tasks.rst

Pattern: Using resources to limit the number of concurrently running tasks
==========================================================================

In this pattern, we use :ref:`resources <resource-requirements>` to limit the number of concurrently running tasks.

Running too many tasks at the same time might overload the cluster and cause issues like OOM.
If that is the case, we can reduce the number of concurrently running tasks by increasing the amount of resources required by those tasks.
This works because Ray makes sure that the sum of the resource requirements of all of the concurrently running tasks on a given node cannot exceed the node's total resources.

Example use case
----------------

You have a data processing workload that processes each input file independently using Ray :ref:`remote functions <ray-remote-functions>`.
Since each task needs to load the input data into heap memory and do the processing, running too many of them can cause OOM.

> In this case, you can use the ``memory`` resource to limit the number of concurrently running tasks (use other resources like ``num_cpus`` can achieve the same goal as well).

Reviewer: ```suggestion
In this case, you can use the ``memory`` resource to limit the number of concurrently running tasks (usage of other resources like ``num_cpus`` can achieve the same goal as well).
```

.. _core-patterns-limit-running-tasks:

Pattern: Using resources to limit the number of concurrently running tasks
==========================================================================

In this pattern, we use :ref:`resources <resource-requirements>` to limit the number of concurrently running tasks.

> Running too many tasks at the same time might overload the cluster and cause issues like OOM.
> If that is the case, we can reduce the number of concurrently running tasks by increasing the amount of resources requested by those tasks.
> This works because Ray makes sure that the sum of the resource requirements of all of the concurrently running tasks on a given node cannot exceed the node's total resources.

Reviewer 2: ```suggestion
By default, Ray tasks require 1 CPU each, so the scheduler limits task concurrrency to the available CPUs.
Tasks that use more than 1 CPU (e.g., via mutlithreading) may experience slowdown due to interference from concurrent tasks, but otherwise are safe to run.

However, tasks that use more than their proportionate share of memory may overload a node and cause issues like OOM.
If that is the case, we can reduce the number of concurrently running tasks by increasing the amount of resources requested by those tasks.
This works because Ray makes sure that the sum of the resource requirements of all of the concurrently running tasks on a given node does not exceed the node's total resources.
```

Also add default number of CPUs for actors? I actually don't know what it is :D

Author: Actor is a bit special, it's 1 cpu for scheduling and 0 cpu for running.

.. note::

   For actor tasks, the number of actors we create controls the number of concurrently running actor tasks we can have.

Example use case
----------------

You have a data processing workload that processes each input file independently using Ray :ref:`remote functions <ray-remote-functions>`.
Since each task needs to load the input data into heap memory and do the processing, running too many of them can cause OOM.

> In this case, you can use the ``memory`` resource to limit the number of concurrently running tasks (usage of other resources like ``num_cpus`` can achieve the same goal as well).

Reviewer 2: Should probably add a note to clarify that memory is a logical resource.

Also, it would be nice to link or mention @clarng's ongoing memory monitor work.

Tasks or actors that use more than 1 CPU (e.g., via mutlithreading) may experience slowdown due to interference from concurrent ones, but otherwise are safe to run.

However, tasks or actors that use more than their proportionate share of memory may overload a node and cause issues like OOM.
If that is the case, we can reduce the number of concurrently running tasks or actors on each node by increasing the amount of resources requested by them.
This works because Ray makes sure that the sum of the resource requirements of all of the concurrently running tasks and actors on a given node does not exceed the node's total resources.

.. note::

   For actor tasks, the number of running actors limits the number of concurrently running actor tasks we can have.

Example use case
----------------

You have a data processing workload that processes each input file independently using Ray :ref:`remote functions <ray-remote-functions>`.
Since each task needs to load the input data into heap memory and do the processing, running too many of them can cause OOM.

> In this case, you can use the logical ``memory`` resource to limit the number of concurrently running tasks (usage of other resources like ``num_cpus`` can achieve the same goal as well).

Reviewer 2: ```suggestion
In this case, you can use the ``memory`` resource to limit the number of concurrently running tasks (usage of other resources like ``num_cpus`` can achieve the same goal as well).
Note that similar to ``num_cpus``, the ``memory`` resource requirement is *logical*, meaning that Ray will not enforce the physical memory usage of each task if it exceeds this amount.
```

.. note::

>    There is active work ongoing

Reviewer 2: Incomplete?

Author: oops. Yea, I was trying to link to @clarng' work but seems that part of doc is not checked in yet. So I'll leave it out for now.
